### PR TITLE
[LIBCLOUD-955] Add DigitalOcean Spaces support.

### PR DESCRIPTION
**Pull request by andrewsomething**
### Description

DigitalOcean Spaces is an S3 compatible object storage service. This adds support for it.

https://developers.digitalocean.com/documentation/spaces/

https://issues.apache.org/jira/browse/LIBCLOUD-955

### Status

- Ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes)

### CODE REVIEW

**Comment by andrewsomething** [comment]
Hmmm... Needs to support Python 2.6. So how are other tests using `assertRaises` as a context manager? :thinking:

**Comment by pquentin** [comment]
Thank you @andrewsomething! Since this is a new driver, we're going to need an [Individual Contributor License Agreement](http://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes) here.

File: docs/storage/_supported_providers.rst
Comment (pquentin) [inline docs/storage/_supported_providers.rst]: I don't know why we did not have trailing spaces before, but can you ensure we only see your changes here by:

1. either generating a new table before adding your changes
1. or keep the trailing spaces as they were before.

Sorry about this, I realize this should be fixed in `generate_provider_feature_matrix_table.py` eventually.

File: libcloud/storage/drivers/digitalocean_spaces.py
Comment (pquentin) [inline libcloud/storage/drivers/digitalocean_spaces.py]: You should only use code from the S3 implementation, not "data", since the S3 driver API version can evolve and break this one. I believe `DEFAULT_SIGNATURE_VERSION` and `API_VERSION` should be defined here. 

You could also choose to define `SIGNATURE_VERSION_V2` and `SIGNATURE_VERSION_V4`: it would avoid using magic numbers in the code but would clutter it too. Up to you.

File: libcloud/storage/drivers/digitalocean_spaces.py
Comment (pquentin) [inline libcloud/storage/drivers/digitalocean_spaces.py]: Can you either use `SIGNATURE_VERSION_V4` (or another similar name) or use `signature_version=4`? (If using a constant forces you to pass a string, that's not a problem.)

File: docs/storage/drivers/digitalocean_spaces.rst
Comment (pquentin) [inline docs/storage/drivers/digitalocean_spaces.rst:1]: Can you please explain how one should choose between v2 signatures or v4 signatures? Maybe point to some documentation?

File: libcloud/storage/drivers/digitalocean_spaces.py
Comment (pquentin) [inline libcloud/storage/drivers/digitalocean_spaces.py:86]: nitpick: Putting the ValueError exception as an `else` branch would avoid repeating '2' and '4'.

File: libcloud/test/storage/test_digitalocean_spaces.py
Comment (pquentin) [inline libcloud/test/storage/test_digitalocean_spaces.py:133]: Can you please add tests for the error conditions?

**Review by pquentin — CHANGES_REQUESTED** [review]
Thanks, the driver looks good! I requested a few changes before we can merge this.

**Comment by andrewsomething** [comment]
@pquentin Thanks for the review! Signed Individual Contributor License Agreement sent to secretary@

While I've got your attention, is there any way for me to get write access to Jira to triage DO-related issues? There are a number that could be closed out. I'm also andrewsomething there.

**Comment by pquentin** [comment]
Thank you, this has been merged into trunk! ✨ 

I reworked the commits a bit to avoid adding commits known to fail the test suite and to better separate the docs changes specific to the driver. I also used `--signoff` per the usual procedure. This is why the PR shows as closed on GitHub. I hope the PR still counts for Hacktoberfest!

Regarding JIRA, I don't have access to it myself, which is why I need to ask you to close LIBCLOUD-955 itself... @Kami could probably help you.

**Comment by andrewsomething** [comment]
Thanks!

**Pull request closed without merging**
